dali/kernels/transpose/transpose_gpu_impl.cuh

```diff
@@ -0,0 +1,248 @@
+    if (threadIdx.x < tile_w) {
+      for (int ty = threadIdx.y, dy = 0; ty < tile_h; ty += blockDim.y, dy += blockDim.y) {
+        #pragma unroll(4)
+        for (int lane = 0; lane < lanes; lane++) {
+          tmp[lane][ty][threadIdx.x] = in[in_ofs + desc.in_strides[ndim-2]*dy + lane];
+        }
+      }
+    }
+    __syncthreads();
+
+    if (threadIdx.x < tile_h) {
+      for (int ty = threadIdx.y, dy = 0; ty < tile_w; ty += blockDim.y, dy += blockDim.y) {
+        #pragma unroll(4)
+        for (int lane = 0; lane < lanes; lane++)
+          out[out_ofs + desc.out_strides[ndim-1]*dy + lane] = tmp[lane][threadIdx.x][ty];
+      }
+    }
+
+    for (int d = ndim - 1; d >= 0; d--) {
+      uint64_t delta = d < ndim-2 ? 1 : kTileSize;  // inner two dimensions are tiled
+      pos[d] += delta;
+      if (pos[d] < desc.shape[d])
+        break;
+      pos[d] = 0;
+    }
+  }
+}
+
+template <typename T>
+__device__ void TransposeTiled(const TiledTransposeDesc<T> &desc) {
+  VALUE_SWITCH(desc.ndim, static_ndim, (2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16),
+    TransposeTiledStatic<static_ndim>(desc),
+    {});
+}
+
+template <typename T>
+__global__ void TransposeTiledSingle(TiledTransposeDesc<T> desc) {
+  TransposeTiled(desc);
+}
+
+template <typename T>
+__global__ void TransposeTiledBatch(const TiledTransposeDesc<T> *descs) {
+  TransposeTiled(descs[blockIdx.y]);
+}
+
+template <typename T>
+struct DeinterleaveDesc {
+  T *__restrict__ out;
+  const T *__restrict__ in;
+
+  fast_div<uint64_t> in_strides[kMaxNDim];
+  uint64_t out_strides[kMaxNDim];
+  uint64_t size;
+  int ndim;
+};
+
+template <typename T>
+__device__ void TransposeDeinterleave(const DeinterleaveDesc<T> &desc) {
+  const int tid = threadIdx.x;
+
+  int ndim = desc.ndim;
+  int lanes = desc.in_strides[ndim-2];
+
+  uint64_t lane_stride = desc.out_strides[ndim-1];
+
+  const uint64_t block_size = blockDim.x;
+  uint64_t start_ofs = (blockIdx.x * block_size + tid) * lanes;
+  uint64_t grid_stride = gridDim.x * block_size * lanes;
+
+  T *out = desc.out;
+  const T *in = desc.in;
+
+  for (uint64_t in_ofs = start_ofs; in_ofs < desc.size; in_ofs += grid_stride) {
+    uint64_t out_ofs = 0;
+    uint64_t tmp_idx = in_ofs;
+    for (int d = 0; d < ndim - 1; d++) {
+      uint64_t a = div_mod(tmp_idx, tmp_idx, desc.in_strides[d]);
+      out_ofs += a * desc.out_strides[d];
+    }
+
+    for (int lane = 0; lane < lanes; lane++) {
+      out[out_ofs + lane * lane_stride] = __ldg(&in[in_ofs + lane]);
+    }
+  }
+}
+
+template <typename T>
+struct GenericTransposeDesc {
+  T *__restrict__ out;
+  const T *__restrict__ in;
+
+  // these are input strides for Deinterleave and output strides for otherwise
+  fast_div<uint64_t> out_strides[kMaxNDim];
+  // ...and vice versa
+  uint64_t in_strides[kMaxNDim];
+  uint64_t size;
+  int ndim;
+};
+
+template <typename T>
+__device__ void TransposeGeneric(const GenericTransposeDesc<T> &desc) {
+  const int tid = threadIdx.x;
```

Comment:
Don't we use this only for the identity case?
If so why do we care about strides inside?

Reply:
Generic is "everything except tiled or deinterleave".